chain-storage/src/lib.rs

```diff
@@ -1,604 +1,424 @@
-            return Err(Error::BlockAlreadyPresent);
-        }
 
-        let block_row_id = tx.last_insert_rowid();
+                if block_info.parent_id() != vec![0; block_info.parent_id().len()].as_slice() {
+                    if info.get(block_info.parent_id())?.is_none() {
+                        return Ok(Err(Error::MissingParent));
+                    }
 
-        let parent = block_info
-            .back_links
-            .iter()
-            .find(|x| x.distance == 1)
-            .unwrap();
+                    let parent_block_info_bin = info.get(block_info.parent_id())?.unwrap();
+                    let mut parent_block_info_reader: &[u8] = &parent_block_info_bin;
+                    let mut parent_block_info =
+                        BlockInfo::deserialize(&mut parent_block_info_reader);
+                    parent_block_info.add_ref();
+                    info.insert(parent_block_info.id(), parent_block_info.serialize())?;
+                }
 
-        let (fast_distance, fast_hash) =
-            match block_info.back_links.iter().find(|x| x.distance != 1) {
-                Some(fast_link) => (
-                    Value::Integer(fast_link.distance as i64),
-                    Value::Blob(fast_link.block_hash.serialize_as_vec().unwrap()),
-                ),
-                None => (Value::Null, Value::Null),
-            };
-
-        tx
-            .prepare_cached("insert into BlockInfo (hash, depth, parent, fast_distance, fast_hash) values(?, ?, ?, ?, ?)")
-            .map_err(|err| Error::BackendError(Box::new(err)))?
-            .execute(&[
-                Value::Blob(block_info.block_hash.serialize_as_vec().unwrap()),
-                Value::Integer(block_info.chain_length as i64),
-                Value::Blob(parent.block_hash.serialize_as_vec().unwrap()),
-                fast_distance,
-                fast_hash,
-            ])
-            .map_err(|err| Error::BackendError(Box::new(err)))?;
+                tips.remove(block_info.parent_id())?;
+                tips.insert(block_info.id(), &[])?;
 
-        let block_info_row_id = tx.last_insert_rowid();
+                let mut height_index = block_info.chain_length().to_le_bytes().to_vec();
+                height_index.extend_from_slice(block_info.id());
+                height_to_block_ids.insert(height_index, &[])?;
 
-        index
-            .add_block(block_info.block_hash.clone(), block_row_id as isize)
-            .unwrap();
+                blocks.insert(block_info.id(), block)?;
 
-        index
-            .add_block_info(
-                block_info.block_hash.clone(),
-                block_info.chain_length,
-                block_info_row_id as isize,
-            )
-            .unwrap();
+                info.insert(block_info.id(), block_info.serialize().as_slice())?;
 
-        tx.commit()
-            .map_err(|err| Error::BackendError(Box::new(err)))?;
+                Ok(Ok(()))
+            },
+        );
```

Comment:
This is a big lambda and could be a free function. It would also allow you to write individual tests eventually.